### PR TITLE
26016 - Feature - Child Landing Zone Management Groups Flexibility

### PR DESCRIPTION
**Pull request by JamJarchitect**
# Overview/Summary

[AB#26268](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/26268)

Update MGDiagAll orchestration and introduction of a parameter to Management Group module that allows the deployer to `true` or `false` whether management groups beneath Platform are wanted in accordance with Trey Research architecture:

https://github.com/Azure/Enterprise-Scale/blob/main/docs/reference/treyresearch/README.md

Replace this with a brief description of what this Pull Request fixes, changes, etc.

## This PR fixes/adds/changes/removes

1. Adds an additional parameter `parPlatformMgAlzDefaultsEnable` to the Management Group module. Values are either `true` or `false` whether child management groups are wanted under Platform management group, with `true` being default
2. Updated `mgDiagsAll.bicep` module to include Platform MG and any custom child MGs under there

### Breaking Changes

1. N/A

## Testing Evidence

image

## As part of this Pull Request I have

- [X] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [X] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [X] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [X] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [X] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [X] Performed testing and provided evidence.
- [X] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [X] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [X] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Great start here @JamJarchitect. However, I think this isn't quite complete yet.

You've nailed the flexibility of enabling/disabling platform children that re defaults (management, connectivity, identity) but you have enabled customers to be able to set their own custom management groups beneath the platform MG, like we say is possible in the tailoring guidance (aka.ms/alz/tailoring)

- [x] Add the ability to create custom MGs beneath Platform MG, like we do on the LZ MG here (https://github.com/JamJarchitect/ALZ-Bicep/blob/26268submggroups/infra-as-code/bicep/modules/managementGroups/managementGroups.bicep#L27-L28)
    - You'll also need the lovely `var` here: https://github.com/JamJarchitect/ALZ-Bicep/blob/26268submggroups/infra-as-code/bicep/modules/managementGroups/managementGroups.bicep#L84 
- [ ] Add the same support into the [`mgDiagSettingsAll.bicep`](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/orchestration/mgDiagSettingsAll/mgDiagSettingsAll.bicep) module

**Comment by JamJarchitect** [comment]
> Great start here @JamJarchitect. However, I think this isn't quite complete yet.
> 
> You've nailed the flexibility of enabling/disabling platform children that re defaults (management, connectivity, identity) but you have enabled customers to be able to set their own custom management groups beneath the platform MG, like we say is possible in the tailoring guidance (aka.ms/alz/tailoring)
> 
> * [x]  Add the ability to create custom MGs beneath Platform MG, like we do on the LZ MG here (https://github.com/JamJarchitect/ALZ-Bicep/blob/26268submggroups/infra-as-code/bicep/modules/managementGroups/managementGroups.bicep#L27-L28)
>   
>   * You'll also need the lovely `var` here: https://github.com/JamJarchitect/ALZ-Bicep/blob/26268submggroups/infra-as-code/bicep/modules/managementGroups/managementGroups.bicep#L84
> * [x]  Add the same support into the [`mgDiagSettingsAll.bicep`](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/orchestration/mgDiagSettingsAll/mgDiagSettingsAll.bicep) module

Amendments done per your request @jtracey93 

- `do not merge` label removed
- Ready for review

**Comment by jtracey93** [comment]
@JamJarchitect really sorry, but can you also add the capability for platform flexibility to this module as well: https://github.com/Azure/ALZ-Bicep/tree/main/infra-as-code/bicep/orchestration/subPlacementAll as part of this PR.

Thanks 👊👊

**Comment by JamJarchitect** [comment]
Tested all modules except for subplacementall - will require additional info on that test.

Results for the MG Deployments:
image

Parameter defaults to normal children MGs of Platform (Identity, Management, Connectivity) with the ability to `false` that. Also the ability to deploy other child MGs under Platform and `true` or `false` the normal `platform` defaults.

MGDiagsAll orchestration deploys successfully including children MGs (custom and default).

Working on SubPlacementAll test and then will remove `do not merge` label.

**Comment by JamJarchitect** [comment]
**UPDATES:**
- `main` merged into my fork and also this branch
- conflicts have been worked through for suffixes
- tested functionality of `ManagementGroup' module, `subplacementall` orchestration and `mgdiagsall` orchestration on custom platform children

image

Checking the READMEs provide key information for new feature and will remove `do not merge` label ready for review

**Comment by baartch** [comment]
@JamJarchitect Thank you very much for your efforts in providing this flexibility! 👏

I just wanted to mention that if it's not already intended to be created, this flexibility would also be needed in _Deployment Flow_ Step 8:
https://github.com/Azure/ALZ-Bicep/tree/main/infra-as-code/bicep/modules/policy/assignments/alzDefaults

**Comment by jtracey93** [comment]
> @JamJarchitect Thank you very much for your efforts in providing this flexibility! 👏
> 
> I just wanted to mention that if it's not already intended to be created, this flexibility would also be needed in _Deployment Flow_ Step 8: [`main`/infra-as-code/bicep/modules/policy/assignments/alzDefaults](https://github.com/Azure/ALZ-Bicep/tree/main/infra-as-code/bicep/modules/policy/assignments/alzDefaults?rgh-link-date=2023-03-09T10%3A08%3A43Z)

Hey @baartch, glad you like this new feature, we will get it reviewed and merged soon.

We actually dont add the flexbility to the `alzDefaultPolicyAssignments` module as its more of a "hardcoded" bicep file that aligns to the default ALZ hierarchy, so no need for flexibility.

If you wanted to make new policy assignments to the newly created management groups via the flexibility input features, you should follow the guidance here: https://github.com/Azure/ALZ-Bicep/wiki/AssigningPolicies

Hope that helps and makes sense

Thanks

Jack

**Comment by stalejohnsen** [comment]
Great addition! Any thoughts on adding flexibility to create MGs below level 3 as well? For example there could be situations where you want to further split corp in isolated and connected sub-archetypes with different guard rails but still inherit the common corp guard rails. Like this module: https://github.com/tyconsulting/BlogPosts/tree/master/Azure-Bicep/management.group

**Comment by jtracey93** [comment]
> Great addition! Any thoughts on adding flexibility to create MGs below level 3 as well? For example there could be situations where you want to further split corp in isolated and connected sub-archetypes with different guard rails but still inherit the common corp guard rails. Like this module: [tyconsulting/BlogPosts@`master`/Azure-Bicep/management.group](https://github.com/tyconsulting/BlogPosts/tree/master/Azure-Bicep/management.group?rgh-link-date=2023-03-17T14%3A38%3A24Z)

Afraid not @stalejohnsen as this goes against our tailoring guidance. Go wide, not deep is our guidance. https://aka.ms/alz/tailoring

**Comment by stalejohnsen** [comment]
> > Great addition! Any thoughts on adding flexibility to create MGs below level 3 as well? For example there could be situations where you want to further split corp in isolated and connected sub-archetypes with different guard rails but still inherit the common corp guard rails. Like this module: [tyconsulting/BlogPosts@`master`/Azure-Bicep/management.group](https://github.com/tyconsulting/BlogPosts/tree/master/Azure-Bicep/management.group?rgh-link-date=2023-03-17T14%3A38%3A24Z)
> 
> Afraid not @stalejohnsen as this goes against our tailoring guidance. Go wide, not deep is our guidance. https://aka.ms/alz/tailoring

Not sure if I follow you there @jtracey93. I would argue that my suggestion of flexibility on the level/layer below corp/online is supported by the following statements in tailoring guide:

* Always try to build on the existing archetypes and hierarchy to meet new requirements. 
If I have a sub-archetype of corp, meaning - I want to differentiate my corp application with slightly different guardrails but still I want to inherit the common corp guard rails. Example: isolated corp applications with no public ips and connected corp applications (like waf mission-critical architecture). If I create a new management group on the same layer/level as corp and online - I don't have the inheritance anymore and instead I'm starting on scratch with defining the guard rail for new archetype, not a related archetype.  

* Avoid going beyond a hierarchy depth of four layers to avoid complexity and unnecessary exclusions.
My suggestion does not go beyond four layers.

**Comment by jtracey93** [comment]
> 

You left out the important part of the last points sentence "Expand archetypes horizontally instead of vertically in the hierarchy." 

Today we have not seen a requirement for this and do no plan to enable this feature, please feel to raise a feature request for this on the repo, with a diagram, and some justification for review 👍👍

**Comment by stalejohnsen** [comment]
> > 
> 
> You left out the important part of the last points sentence "Expand archetypes horizontally instead of vertically in the hierarchy."
> 
> Today we have not seen a requirement for this and do no plan to enable this feature, please feel to raise a feature request for this on the repo, with a diagram, and some justification for review 👍👍
 
Will add this as a new feature request, thanks

File: infra-as-code/bicep/modules/managementGroups/managementGroups.bicep [hunk 1]
```diff
@@ -159,52 +163,27 @@ resource resDecommissionedMg 'Microsoft.Management/managementGroups@2021-04-01'
...
-}
-
-// Level 3 - Child Management Groups under Landing Zones MG
+}]
 
-resource resLandingZonesChildMgs 'Microsoft.Management/managementGroups@2021-04-01' = [for mg in items(varLandingZoneMgChildrenUnioned): if (!empty(varLandingZoneMgChildrenUnioned)) {
-  name: '${parTopLevelManagementGroupPrefix}-landingzones-${mg.key}${parTopLevelManagementGroupSuffix}'
+//Level 3 - Child Management Groups under Platform MG
+resource resPlatformChildMgs 'Microsoft.Management/managementGroups@2021-04-01' = [for mg in items(varPlatformMgChildrenUnioned): if (!empty(varPlatformMgChildrenUnioned)) {
+  name: '${parTopLevelManagementGroupPrefix}-platform-${mg.key}'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:181]:
```suggestion
  name: '${parTopLevelManagementGroupPrefix}-platform-${mg.key}${parTopLevelManagementGroupSuffix}'
```

Think we are missing the suffix additions

File: infra-as-code/bicep/modules/managementGroups/managementGroups.bicep [hunk 2]
```diff
@@ -221,9 +200,8 @@ module modCustomerUsageAttribution '../../CRML/customerUsageAttribution/cuaIdTen
 output outTopLevelManagementGroupId string = resTopLevelMg.id
 
 output outPlatformManagementGroupId string = resPlatformMg.id
-output outPlatformManagementManagementGroupId string = resPlatformManagementMg.id
-output outPlatformConnectivityManagementGroupId string = resPlatformConnectivityMg.id
-output outPlatformIdentityManagementGroupId string = resPlatformIdentityMg.id
+output outPlatformChildrenManagementGroupIds array = [for mg in items(varLandingZoneMgChildrenUnioned): '/providers/Microsoft.Management/managementGroups/${parTopLevelManagementGroupPrefix}-platform-${mg.key}']
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:203]:
```suggestion
output outPlatformChildrenManagementGroupIds array = [for mg in items(varLandingZoneMgChildrenUnioned): '/providers/Microsoft.Management/managementGroups/${parTopLevelManagementGroupPrefix}-platform-${mg.key}${parTopLevelManagementGroupSuffix}']
```

File: infra-as-code/bicep/orchestration/mgDiagSettingsAll/mgDiagSettingsAll.bicep [hunk 3]
```diff
@@ -55,9 +65,17 @@ var varLandingZoneMgCustomChildren = [for customMg in parLandingZoneMgChildren:
   mgId: '${parTopLevelManagementGroupPrefix}-landingzones-${customMg}${parTopLevelManagementGroupSuffix}'
 }]
 
+// Used if parLandingZoneMgConfidentialEnable not empty
+var varPlatformMgCustomChildren = [for customMg in parPlatformMgChildren: {
+  mgId: '${parTopLevelManagementGroupPrefix}-platform-${customMg}'
```
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/mgDiagSettingsAll/mgDiagSettingsAll.bicep:70]:
```suggestion
  mgId: '${parTopLevelManagementGroupPrefix}-platform-${customMg}${parTopLevelManagementGroupSuffix}'
```

File: infra-as-code/bicep/orchestration/mgDiagSettingsAll/parameters/mgDiagSettingsAll.parameters.all.json [hunk 4]
```diff
@@ -11,15 +11,21 @@
...
     },
+    "parPlatformMgAlzDefaultsEnable": {
+      "value": true
+    },
     "parLandingZoneMgConfidentialEnable": {
       "value": false
     },
     "parLogAnalyticsWorkspaceResourceId": {
-      "value": "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourcegroups/alz-logging/providers/microsoft.operationalinsights/workspaces/alz-log-analytics"
+      "value": "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourcegroups/logging-rg/providers/microsoft.operationalinsights/workspaces/alz-log-analytics"
```
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/mgDiagSettingsAll/parameters/mgDiagSettingsAll.parameters.all.json:21]: why has this changed?

File: infra-as-code/bicep/orchestration/subPlacementAll/subPlacementAll.bicep [hunk 5]
```diff
@@ -133,6 +136,16 @@ module modplatformIdentityMgSubPlacement '../../modules/subscriptionPlacement/su
   }
 }
 
+// Custom Children Platform Management Groups
+module modPlatformMgChildrenSubPlacement '../../modules/subscriptionPlacement/subscriptionPlacement.bicep' = [for mg in items(parPlatformMgChildrenSubs): if (!empty(parPlatformMgChildrenSubs)) {
+  name: take('modPlatformMgChildrenSubPlacement-${uniqueString(mg.key, string(length(mg.value.subscriptions)), deployment().name)}', 64)
+  scope: managementGroup('${parTopLevelManagementGroupPrefix}-platform-${mg.key}')
```
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/subPlacementAll/subPlacementAll.bicep:142]:
```suggestion
  scope: managementGroup('${parTopLevelManagementGroupPrefix}-platform-${mg.key}${parTopLevelManagementGroupSuffix}')
```

File: infra-as-code/bicep/orchestration/subPlacementAll/subPlacementAll.bicep [hunk 6]
```diff
@@ -133,6 +136,16 @@ module modplatformIdentityMgSubPlacement '../../modules/subscriptionPlacement/su
...
+  params: {
+    parTargetManagementGroupId: '${parTopLevelManagementGroupPrefix}-platform-${mg.key}'
```
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/subPlacementAll/subPlacementAll.bicep:144]:
```suggestion
    parTargetManagementGroupId: '${parTopLevelManagementGroupPrefix}-platform-${mg.key}${parTopLevelManagementGroupSuffix}'
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Great work here @JamJarchitect i think we are just missing the MG suffix work that @mbrat2005 added in #462 in this PR.

I think i have added suggestion for everywhere they are needed, please review, accept and commit them if so 👍

**Comment by jtracey93** [comment]
stil active

File: infra-as-code/bicep/orchestration/mgDiagSettingsAll/parameters/mgDiagSettingsAll.parameters.all.json (hunk 4 above)
Comment (JamJarchitect) [reply infra-as-code/bicep/orchestration/mgDiagSettingsAll/parameters/mgDiagSettingsAll.parameters.all.json]: changed this back

**Pull request closed without merging**